### ISSUE TITLE
Retrieve skeleton tree structure

### ISSUE THREAD

**Issue by constantinpape** [jni/skan#70]
Skeletons are supposed to be trees (though we could give `skan` a loopy input image. I don't know what happens then yet, but that's stuff for a different issue).
It would be nice to have some functionality to retrieve this, i.e. have some function
that returns the parent node for each skeleton node.
This could also be encoded in the node-ids, e.g. by the parent node always having the smaller id,
like in this toy example:
```
input image
[[0 0 0 1 0 0 0]
 [0 0 0 1 0 0 0]
 [0 0 0 1 0 0 0]
 [1 1 1 1 1 1 1]
 [0 0 0 1 0 0 0]
 [0 0 0 1 0 0 0]
 [0 0 0 1 0 0 0]]
skan node labeling (yes there is a bug, see below)
[[ 0  0  0  1  0  0  0]
 [ 0  0  0  2  0  0  0]
 [ 0  0  0  0  0  0  0]
 [ 4  5  6  7  8  9 10]
 [ 0  0  0 11  0  0  0]
 [ 0  0  0 12  0  0  0]
 [ 0  0  0 13  0  0  0]]
tree node labeling (one option)
[[ 0  0  0  7  0  0  0]
 [ 0  0  0  6  0  0  0]
 [ 0  0  0  5  0  0  0]
 [ 4  3  2  1  8  9 10]
 [ 0  0  0 11  0  0  0]
 [ 0  0  0 12  0  0  0]
 [ 0  0  0 13  0  0  0]]
```
Note that there is a bug in the skan node labeling (what should be 3 is a 0). I am using 0.7.1 though, which is the latest version on pypi (I can't use conda yet due to https://github.com/conda-forge/skan-feedstock/issues/12), so maybe this is fixed in 0.8.1.

Finally, for my use case, this is motivated by the need to serialize to [.swc](http://research.mssm.edu/cnic/swc.html), which expects a tree representation.

**Comment by jni** [comment]
No, skeletons aren't necessarily trees, they are trees in one particular domain. ;) See the [skan paper](https://peerj.com/articles/4312/) for  a non-tree application. =) It might be possible to use scipy.sparse.csgraph to get a tree labelling though, if you know the root.

The bug is concerning and might be due to the junction disambiguation code (see the suppl figure in the paper). Can you let me know if you still see this in master?

**Comment by constantinpape** [comment]
> they are trees in one particular domain

You're right, this is only true in the narrow-minded connectomics sense :).

I will implement some functions to check if a skeleton is a tree or loopy (or do you have that already in `skan`?) and get a tree labeling (will check if I can use scipy.sparse).
We can then see if you are interested to have this as a contribution.

> See the [skan paper](https://peerj.com/articles/4312/)

Thanks for the link, I will check out the paper.


> Can you let me know if you still see this in master?

I will check later and let you know.

**Comment by jni** [comment]
Ah. I see. (btw you mean (2, 3) ;) 

Anyway, nope, still the same issue, sorry that this is a bit confusing but it's still "correct", based on the junction uniquifying that skan does. Basically, the junction made up of nodes `{3, 6, 7, 8, 11}` gets condensed into the smallest-index node (3), and the coordinate of that junction becomes the centroid of the nodes within it. The other values are dropped out of the graph. See thusly:

```python
In [7]: g, coord, deg = csr.skeleton_to_csgraph(x)
In [15]: list(set(map(frozenset, g.todok().keys())))
Out[15]: 
[frozenset({1, 2}),
 frozenset({3, 12}),
 frozenset({3, 5}),
 frozenset({12, 13}),
 frozenset({3, 9}),
 frozenset({2, 3}),
 frozenset({9, 10}),
 frozenset({4, 5})]
```

The old, now-unused coordinates stay there because, well, I was lazy I guess?

If you want to turn off this behaviour, you can use `unique_junctions=False`:

```python
In [8]: g2, coord2, deg2 = csr.skeleton_to_csgraph(x, unique_junctions=False)
In [9]: coord2[3]
Out[9]: array([2., 3.])
```

This all should be at least documented, that much is certain!

**Comment by jni** [comment]
@constantinpape sorry if I sounded a bit dismissive! Can you tell me if that flag solves the problem for you? Specifically, does swc require that each pixel in a skeleton be represented? Or can it use "real-space" coordinates for points? ie, can I do anything to make sure that you can use skan while still maintaining the functionality described above?

**Comment by constantinpape** [comment]
> sorry if I sounded a bit dismissive!

No worries, I was just distracted from this by other projects. Hope to work more on the skeleton related things soon.

> Can you tell me if that flag solves the problem for you?

Yes, it did.

> Specifically, does swc require that each pixel in a skeleton be represented? Or can it use "real-space" coordinates for points?

Yes, .swc can use "real-space" coordinates. In most cases the coordinates are transformed to the phyiscal units (e.g. nanometer) anyway.

>  ie, can I do anything to make sure that you can use skan while still maintaining the functionality described above?

The only thing that is missing is going from the graph representation to a tree representation (i.e. a representation where each node gets assigned a parent node). As you have correctly remarked, skeleton-graphs don't have tree structure in general, but it should still be possible to get the parent-ids.

Anyway, this is pretty generic graph processing and doesn't necessary need to be in skan.
I have started to implement a [cycle check with depth-first search](https://github.com/constantinpape/skanner/blob/master/skanner/graph.py) for the skan graph and it should be easy to adapt this to get the parent ids.
